### PR TITLE
Add `WithConsoleUrl` interface

### PR DESCRIPTION
**Pull request by Vlatombe**
This simplifies the logic of `Functions#getConsoleUrl` and makes it pluggable.

This also clarifies `ConsoleUrlProvider#getConsoleUrl` and make it a requirement for implementations to return relative url w/o leading slash.

### Testing done

Checked known implementations (OSS + CloudBees)
* [x] [pipeline-graph-view](https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/fe92778e4a0e550cc8acb1df007d0555d2fef24e/src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewUrlProvider.java#L17-L22)
* [x] cloudbees-pipeline-explorer

### Proposed changelog entries

- N/A

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [X] The Jira issue, if it exists, is well-described.
- [X] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/jenkins/console/ConsoleUrlProvider.java
Comment (dwnusbaum) [inline core/src/main/java/jenkins/console/ConsoleUrlProvider.java:66]: It would be nice to fix this typo:

```suggestion
     * if this implementation does not want to serve a special console view for this build.
```

File: core/src/main/java/hudson/model/Queue.java
Comment (dwnusbaum) [inline core/src/main/java/hudson/model/Queue.java:2136]: Did you test this case to make sure it works?

File: core/src/main/java/hudson/Functions.java
Comment (dwnusbaum) [inline core/src/main/java/hudson/Functions.java:1915]: Can this be adjusted to accept the new interface instead? Maybe also worth adjusting the Javadoc to replace mentions of `executable` with something more generic.

**Review by dwnusbaum — APPROVED** [review]

File: core/src/main/java/hudson/Functions.java
Comment (Vlatombe) [reply core/src/main/java/hudson/Functions.java]: done

**Review by dwnusbaum — APPROVED** [review]

File: core/src/main/java/hudson/model/Queue.java
Comment (Vlatombe) [reply core/src/main/java/hudson/model/Queue.java:2136]: This is actually moved code from `Functions#getConsoleUrl` but I doublechecked to make sure.

File: core/src/main/java/hudson/Functions.java
Comment (StefanSpieker) [inline core/src/main/java/hudson/Functions.java:1911]: After the last rework it isn't an object any longer, maybe also adjust the javadoc here

Comment (jglick) [reply core/src/main/java/hudson/Functions.java:1911]: Well, a “model object”. Seems OK to me.

**Review by jglick — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Comment by Vlatombe** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
